pkg/neonvm/ipam/ipam.go

```diff
@@ -168,174 +229,84 @@ func LoadFromNad(nadConfig string, nadNamespace string) (*IPAMConfig, error) {
 	return n.IPAM, nil
 }
 
-// Performing IPAM actions with Leader Election to avoid duplicates
-func (i *IPAM) acquireORrelease(ctx context.Context, vmName string, vmNamespace string, action int) (net.IPNet, error) {
-	var ip net.IPNet
-	var err error
-	var ipamerr error
-
-	leOverallTimeout := IpamRequestTimeout * 3
-	lockName := "neonvmipam"
-	lockIdentity := fmt.Sprintf("%s/%s", vmNamespace, vmName)
-
-	// define resource lock
-	lock := &resourcelock.LeaseLock{
-		LeaseMeta: metav1.ObjectMeta{
-			Name:      lockName,
-			Namespace: i.Config.NetworkNamespace,
-		},
-		Client: i.kubeClient.CoordinationV1(),
-		LockConfig: resourcelock.ResourceLockConfig{
-			Identity: lockIdentity,
-		},
-	}
-
-	done := make(chan struct{})
-
-	// define leader elector
-	le, err := leaderelection.NewLeaderElector(leaderelection.LeaderElectionConfig{
-		Lock:            lock,
-		ReleaseOnCancel: true,
-		LeaseDuration:   time.Millisecond * time.Duration(DefaultLeaderLeaseDurationMs),
-		RenewDeadline:   time.Millisecond * time.Duration(DefaultLeaderRenewDeadlineMs),
-		RetryPeriod:     time.Millisecond * time.Duration(DefaultLeaderRetryPeriodMs),
-		Callbacks: leaderelection.LeaderCallbacks{
-			OnStartedLeading: func(c context.Context) {
-				ip, ipamerr = i.runIPAM(ctx, vmName, vmNamespace, action)
-				close(done)
-				<-c.Done()
-			},
-			OnStoppedLeading: func() {
-				// do nothing
-				err = error(nil)
-			},
-		},
-	})
-	if err != nil {
-		return ip, err
-	}
-
-	// context with timeout for leader elector
-	leCtx, leCancel := context.WithTimeout(ctx, leOverallTimeout)
-	defer leCancel()
-
-	// run election in background
-	var wg sync.WaitGroup
-	wg.Add(1)
-	go func() { defer wg.Done(); le.Run(leCtx) }()
-
-	// wait until job was done and then cancel election context
-	// or exit with error when context got timeout
-	select {
-	case <-done:
-		leCancel()
-	case <-leCtx.Done():
-		err = errors.New("context got timeout while waiting to become leader")
-	}
-	if err != nil {
-		return ip, err
-	}
-
-	wg.Wait()
-
-	// ip.String() returns string "<nil>" on errors in ip struct parsing or if *ip is nil
-	if ip.String() == "<nil>" {
-		return ip, errors.New("something wrong, probably with leader election")
-	}
-
-	return ip, ipamerr
-}
-
 // Performing IPAM actions
-func (i *IPAM) runIPAM(ctx context.Context, vmName string, vmNamespace string, action int) (net.IPNet, error) {
+func (i *IPAM) runIPAM(ctx context.Context, action ipamAction) (net.IPNet, error) {
+	var err error
 	var ip net.IPNet
-	var ipamerr error
+	log := log.FromContext(ctx)
 
-	// check action
-	switch action {
-	case Acquire, Release:
-	default:
-		return ip, fmt.Errorf("got an unknown action: %v", action)
+	ok := i.concurrencyLimiter.TryAcquire(1)
+	if !ok {
+		return net.IPNet{}, ErrAgain
```

Comment:
suggestion: Can you add a comment here explaining why we have this concurrency limiter _and_ the mutex below it?

Reply:
Done.